README.md
# ⛔️ DEPRECATED

Tailwind CSS 1.2 (released in February 2020) has utilities for transition [property](https://tailwindcss.com/docs/transition-property), [duration](https://tailwindcss.com/docs/transition-duration), and [timing function](https://tailwindcss.com/docs/transition-timing-function). It doesn’t have transition delay or `will-change` utilities (yet), nor does it have a way to define a default transition duration or timing function, but I feel like keeping this plugin around would be more confusing than helpful since it uses similar (or identical in some cases) class names as the core transition utilities. Feel free to fork it if you want, but I recommend migrating to Tailwind’s official solution and write custom utilities for `transition-delay` and `will-change` when needed.

# Transitions Plugin for Tailwind CSS

## Installation

```bash
npm install tailwindcss-transitions
```

## Usage

```js
// tailwind.config.js
module.exports = {
  theme: {
    transitionProperty: { // defaults to these values
      'none': 'none',
      'all': 'all',
      'default': ['background-color', 'border-color', 'color', 'fill', 'stroke', 'opacity', 'box-shadow', 'transform'],
      'colors': ['background-color', 'border-color', 'color', 'fill', 'stroke'],
      'bg': 'background-color',
      'border': 'border-color',
      'color': 'color',
      'opacity': 'opacity',
      'shadow': 'box-shadow',
      'transform': 'transform',
    },
    transitionDuration: { // defaults to these values
      'default': '250ms',
      '0': '0ms',
      '50': '50ms',
      '75': '75ms',
      '100': '100ms',
      '150': '150ms',
      '200': '200ms',
      '250': '250ms',
      '300': '300ms',
      '400': '400ms',
      '500': '500ms',
      '750': '750ms',
      '1000': '1000ms',
    },
    transitionTimingFunction: { // defaults to these values
      'default': 'ease',
      'linear': 'linear',
      'ease': 'ease',
      'ease-in': 'ease-in',
      'ease-out': 'ease-out',
      'ease-in-out': 'ease-in-out',
    },
    transitionDelay: { // defaults to these values
      'default': '0ms',
      '0': '0ms',
      '100': '100ms',
      '250': '250ms',
      '500': '500ms',
      '750': '750ms',
      '1000': '1000ms',
    },
    willChange: { // defaults to these values
      'auto': 'auto',
      'scroll': 'scroll-position',
      'contents': 'contents',
      'opacity': 'opacity',
      'transform': 'transform',
    },
  },
  variants: { // all the following default to ['responsive']
    transitionProperty: ['responsive'],
    transitionDuration: ['responsive'],
    transitionTimingFunction: ['responsive'],
    transitionDelay: ['responsive'],
    willChange: ['responsive'],
  },
  plugins: [
    require('tailwindcss-transitions')(),
  ],
};
```

The default configuration generates the following CSS:

```css
/* base styles for the default transition duration, timing function, and delay (when they differ from the CSS defaults) */
*, *::before, *::after {
  --transition-duration: 250ms;
  /* when the default timing function is a value other than "ease": */
  --transition-timing-function: [default-timing-function];
  /* when the default delay is a value other than zero: */
  --transition-delay: [default-delay];
}

/* configurable with the "transitionProperty" theme object */
.transition-none {
  transition-property: none;
  transition-duration: 250ms;
  transition-duration: var(--transition-duration);
}
.transition-all {
  transition-property: all;
  transition-duration: 250ms;
  transition-duration: var(--transition-duration);
}
.transition {
  transition-property: background-color, border-color, color, fill, stroke, opacity, box-shadow, transform;
  transition-duration: 250ms;
  transition-duration: var(--transition-duration);
}
.transition-[property-key] {
  transition-property: [property-value];
  /* when the default duration is a value other than zero: */
  transition-duration: [default-duration];
  transition-duration: var(--transition-duration);
  /* when the default timing function is a value other than "ease": */
  transition-timing-function: [default-timing-function];
  transition-timing-function: var(--transition-timing-function);
  /* when the default delay is a value other than zero: */
  transition-delay: [default-delay];
  transition-delay: var(--transition-delay);
}

/* configurable with the "transitionDuration" theme object */
.transition-0 {
  --transition-duration: 0ms;
  transition-duration: 0ms;
  transition-duration: var(--transition-duration);
}
.transition-50 {
  --transition-duration: 50ms;
  transition-duration: 50ms;
  transition-duration: var(--transition-duration);
}
.transition-[duration-key] {
  transition-duration: [duration-value];
  /* when the default duration is a value other than zero: */
  --transition-duration: [duration-value];
  transition-duration: var(--transition-duration);
  /* when the default timing function is a value other than "ease": */
  transition-timing-function: [default-timing-function];
  transition-timing-function: var(--transition-timing-function);
  /* when the default delay is a value other than zero: */
  transition-delay: [default-delay];
  transition-delay: var(--transition-delay);
}

/* configurable with the "transitionTimingFunction" theme object */
.transition-linear {
  transition-timing-function: linear;
}
.transition-ease {
  transition-timing-function: ease;
}
.transition-[timing-function-key] {
  transition-timing-function: [timing-function-value];
  /* when the default timing function is a value other than "ease": */
  --transition-timing-function: [timing-function-value];
  transition-timing-function: var(--transition-timing-function);
}

/* configurable with the "transitionDelay" theme object */
.transition-delay-0 {
  transition-delay: 0ms;
}
.transition-delay-100 {
  transition-delay: 100ms;
}
.transition-delay-[key] {
  transition-delay: [value];
  /* when the default delay is a value other than zero: */
  --transition-delay: [value];
  transition-delay: var(--transition-delay);
}

/* configurable with the "willChange" theme object */
.will-change {
  will-change: contents;
}
.will-change-auto {
  will-change: auto;
}
.will-change-[key] {
  will-change: [value];
}
```

Which you can then use in your HTML like this:

```html
<button class="bg-gray-600 hover:bg-gray-500 transition-bg">
  Hover me for a lighter background
</button>

<button class="bg-gray-200 hover:bg-gray-900 text-gray-900 hover:text-gray-200 transition-colors transition-500 transition-linear">
  Hover me to invert colors
</button>
```

Note: The `transitionProperty`, `transitionDuration`, `transitionTimingFunction`, and `transitionDelay` theme objects accept a `default` key. For `transitionProperty`, it generates a simple `transition` class (instead of `transition-default`), but for the other three, `default` doesn’t generate any class; it is used to define a custom property on all elements and pseudo-elements (`*, *::before, *::after`) if the value differs from the CSS-defined default. These custom properties are then used to set actual properties on elements that have a `transition-[property]` or `transition-[duration]` class, so that you don’t have to define a duration, timing function, or delay every time.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project mostly adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [2.2.0] - 2020-02-03

### Added
- Added a `default` transition property, which generates a simple `transition` class and includes all the properties from `colors` as well as `opacity`, `box-shadow`, and `transform` (taken from Tailwind 1.2)
- Added a `shadow` transition property (taken from Tailwind 1.2)
- Added `50`, `75`, `150`, `200`, `300`, and `400` transition durations

### Changed
- The `transition-colors` utility now includes the `fill` and `stroke` properties, in addition to `background-color`, `border-color`, and `color` (taken from Tailwind 1.2)

## [2.1.0] - 2019-09-04

### Changed since 2.1.0-beta.2
- Allowed for durations and delays to be defined as numbers but be output as `ms`

### Changed since 2.0.x
- Changed the implementation of the `default` duration, timing function, and delay to prevent generating base styles that change the value of `transition-property`, `transition-duration`, `transition-timing-function`, and `transition-delay` on all elements, in order to prevent conflicts with libraries that don’t expect these properties to differ from their CSS-defined default
- Base styles targeting all elements are still generated if a `default` duration, timing function, and/or delay is set (which is the case by default: the `default` duration is `250ms`), but they define custom properties that are very unlikely to conflict with third-party libraries (e.g. `--transition-duration`); these are then used to set actual properties on elements that have a `transition-[property]` or `transition-[duration]` class
- As a result, it is now possible to only use a transition duration utility (e.g. `transition-500`) on an element to make it transition all its properties; no need for `transition-all` anymore, since the default value of `transition-property` is already `all`

## [2.1.0-beta.2] - 2019-09-01

**Release `2.1.0-beta.1` was deprecated and reverted, so the following changes were made from release `2.0.1`:**

### Changed
- Changed the implementation of the `default` duration, timing function, and delay to prevent generating base styles that change the value of `transition-property`, `transition-duration`, `transition-timing-function`, and `transition-delay` on all elements, in order to prevent conflicts with libraries that don’t expect these properties to differ from their CSS-defined default
- Base styles targeting all elements are still generated if a `default` duration, timing function, and/or delay is set (which is the case by default: the `default` duration is `250ms`), but they define custom properties that are very unlikely to conflict with third-party libraries (e.g. `--transition-duration`); these are then used to set actual properties on elements that have a `transition-[property]` or `transition-[duration]` class
- As a result, it is now possible to only use a transition duration utility (e.g. `transition-500`) on an element to make it transition all its properties; no need for `transition-all` anymore, since the default value of `transition-property` is already `all`

## [2.1.0-beta.1] - 2019-08-01 [DEPRECATED]

**Decided to go in another direction for v2.1, and reverted most of the changes in this release.**

### Added
- Added the ability to set a `default` transition property, which is set to `auto` by default, meaning that it depends on the `default` transition duration: if it’s `0s` or `0ms` then `auto` resolves to `all`, otherwise it resolves to `none`

### Changed
- Changed the `default` transition duration from `250ms` to `0ms`
- As a result of the above, base styles targeting all elements are no longer generated by default since the default property and duration are now the same as the CSS defaults (`all` and `0s`, respectively)
- Therefore, with the new defaults you only need a transition duration utility (e.g. `transition-fast`) on an element to make it transition all its properties (no need for `transition-all`); if you want the previous behaviour, simply set the `default` transition duration to a non-zero value
- Changed the default transition durations from a numeric scale (`250`, `500`, `750`, etc.) to a descriptive scale (`slow`, `medium`, `fast`, etc.)
- Changed the default transition delays to match the new duration values, but with a numeric scale (`0`, `200`, `400`, `600`, `800`, and `1000`)

## [2.0.1] - 2019-05-16

### Changed
- The base styles which set the default duration, timing function, and delay are now applied to pseudo-elements as well (changed the `*` selector to `*, *::before, *::after`)

## [2.0.0] - 2019-05-13

### Changed since 2.0.0-beta.2
- Added support for global variants thanks to Tailwind’s `variants()` helper function

### Added since 1.x
- Tailwind 1.0.0 compatibility

### Changed since 1.x
- The plugin doesn’t accept a config object anymore; instead it finds what it needs in the `theme` and `variants` keys of your config (see `README` for more info)
- Multiple utilities as well as responsive variants are now generated by default (see `README` for more info)
- Shorter class names for transition durations (`transition-duration-500` is now `transition-500`) and timing functions (`transition-timing-ease` is now `transition-ease`)
- Transition utilities now set `transition-property` instead of the `transition` shorthand; as a result, responsive transition utilities don’t reset the duration, timing function or delay anymore
- For the above to work, the plugin now sets some base styles on all elements (changes `transition-property`’s default value of `all` to `none`, and sets the default duration, timing function, and delay defined in the theme)

## [2.0.0-beta.2] - 2019-04-07

### Removed

index.js
const _ = require('lodash');

const time = time => _.isNumber(time) ? `${time}ms` : time;

module.exports = function() {
  return ({ theme, variants, e, addBase, addUtilities }) => {
    const defaultPropertyTheme = {
      'none': 'none',
      'all': 'all',
      'default': ['background-color', 'border-color', 'color', 'fill', 'stroke', 'opacity', 'box-shadow', 'transform'],
      'colors': ['background-color', 'border-color', 'color', 'fill', 'stroke'],
      'bg': 'background-color',
      'border': 'border-color',
      'color': 'color',
      'opacity': 'opacity',
      'shadow': 'box-shadow',
      'transform': 'transform',
    };
    const defaultPropertyVariants = ['responsive'];
    const defaultDurationTheme = {
      'default': '250ms',
      '0': '0ms',
      '50': '50ms',
      '75': '75ms',
      '100': '100ms',
      '150': '150ms',
      '200': '200ms',
      '250': '250ms',
      '300': '300ms',
      '400': '400ms',
      '500': '500ms',
      '750': '750ms',
      '1000': '1000ms',
    };
    const defaultDurationVariants = ['responsive'];
    const defaultTimingFunctionTheme = {
      'default': 'ease',
      'linear': 'linear',
      'ease': 'ease',
      'ease-in': 'ease-in',
      'ease-out': 'ease-out',
      'ease-in-out': 'ease-in-out',
    };
    const defaultTimingFunctionVariants = ['responsive'];
    const defaultDelayTheme = {
      'default': '0ms',
      '0': '0ms',
      '100': '100ms',
      '250': '250ms',
      '500': '500ms',
      '750': '750ms',
      '1000': '1000ms',
    };
    const defaultDelayVariants = ['responsive'];
    const defaultWillChangeTheme = {
      'auto': 'auto',
      'scroll': 'scroll-position',
      'contents': 'contents',
      'opacity': 'opacity',
      'transform': 'transform',
    };
    const defaultWillChangeVariants = ['responsive'];

    const propertyTheme = theme('transitionProperty', defaultPropertyTheme);
    const propertyVariants = variants('transitionProperty', defaultPropertyVariants);
    const durationTheme = theme('transitionDuration', defaultDurationTheme);
    const durationVariants = variants('transitionDuration', defaultDurationVariants);
    const timingFunctionTheme = theme('transitionTimingFunction', defaultTimingFunctionTheme);
    const timingFunctionVariants = variants('transitionTimingFunction', defaultTimingFunctionVariants);
    const delayTheme = theme('transitionDelay', defaultDelayTheme);
    const delayVariants = variants('transitionDelay', defaultDelayVariants);
    const willChangeTheme = theme('willChange', defaultWillChangeTheme);
    const willChangeVariants = variants('willChange', defaultWillChangeVariants);

    const defaultDuration = time(_.defaults({}, durationTheme, defaultDurationTheme).default);
    const defaultTimingFunction = _.defaults({}, timingFunctionTheme, defaultTimingFunctionTheme).default;
    const defaultDelay = time(_.defaults({}, delayTheme, defaultDelayTheme).default);

    const baseDuration = _.includes(['0', '0s', '0ms'], defaultDuration) ? null : defaultDuration;
    const baseTimingFunction = defaultTimingFunction === 'ease' ? null : defaultTimingFunction;
    const baseDelay = _.includes(['0', '0s', '0ms'], defaultDelay) ? null : defaultDelay;

    const baseStyles = {
      ...(function() {
        if (baseDuration === null && baseTimingFunction === null && baseDelay === null) {
          return {};
        }
        return {
          '*, *::before, *::after': {
            '--transition-duration': baseDuration,
            '--transition-timing-function': baseTimingFunction,
            '--transition-delay': baseDelay,
          }
        };
      })(),
    };

    const durationStyles = value => {
      if (baseDuration === null) {
        return {
          transitionDuration: time(value),
        };
      }
      return {
        '--transition-duration': time(value),
        transitionDuration: [time(value), 'var(--transition-duration)'],
      };
    };

    const timingFunctionStyles = value => {
      if (baseTimingFunction === null) {
        return {
          transitionTimingFunction: value,
        };
      }
      return {
        '--transition-timing-function': value,
        transitionTimingFunction: [value, 'var(--transition-timing-function)'],
      };
    };

    const delayStyles = value => {
      if (baseDelay === null) {
        return {
          transitionDelay: time(value),
        };
      }
      return {
        '--transition-delay': time(value),
        transitionDelay: [time(value), 'var(--transition-delay)'],
      };
    };

    const defaultDurationStyles = {
      transitionDuration: baseDuration === null ? null : [baseDuration, 'var(--transition-duration)'],
    };

    const defaultTimingFunctionStyles = {
      transitionTimingFunction: baseTimingFunction === null ? null : [baseTimingFunction, 'var(--transition-timing-function)'],
    };

    const defaultDelayStyles = {
      transitionDelay: baseDelay === null ? null : [baseDelay, 'var(--transition-delay)'],
    };

    const propertyUtilities = _.fromPairs(
      _.map(propertyTheme, (value, modifier) => {
        return [
          `.${e(`transition${modifier === 'default' ? '' : `-${modifier}`}`)}`,
          {
            transitionProperty: _.isArray(value) ? value.join(', ') : value,
            ...defaultDurationStyles,
            ...defaultTimingFunctionStyles,
            ...defaultDelayStyles,
          },
        ];
      })
    );

    const durationUtilities = _.fromPairs(
      _.map(durationTheme, (value, modifier) => {
        if (modifier === 'default') {
          return [];
        }
        return [
          `.${e(`transition-${modifier}`)}`,
          {
            ...durationStyles(value),
            ...defaultTimingFunctionStyles,
            ...defaultDelayStyles,
          },
        ];
      })
    );

    const timingFunctionUtilities = _.fromPairs(
      _.map(timingFunctionTheme, (value, modifier) => {
        if (modifier === 'default') {
          return [];
        }
        return [
          `.${e(`transition-${modifier}`)}`,
          {
            ...timingFunctionStyles(value),
          },
        ];
      })
    );

    const delayUtilities = _.fromPairs(
      _.map(delayTheme, (value, modifier) => {
        if (modifier === 'default') {
          return [];
        }
        return [
          `.${e(`transition-delay-${modifier}`)}`,
          {
            ...delayStyles(value),
          },
        ];
      })
    );

    const willChangeUtilities = _.fromPairs(
      _.map(willChangeTheme, (value, modifier) => {
        return [
          `.${e(`will-change-${modifier}`)}`,
          {
            willChange: value,
          },
        ];
      })
    );

    addBase(baseStyles);
    addUtilities(propertyUtilities, propertyVariants);
    addUtilities(durationUtilities, durationVariants);
    addUtilities(timingFunctionUtilities, timingFunctionVariants);
    addUtilities(delayUtilities, delayVariants);
    addUtilities(willChangeUtilities, willChangeVariants);
  };
};
